### DISCUSSION TITLE
Library speed when inserting large amounts of RawContacts

### DISCUSSION THREAD

**Discussion by marrale** [vestrel00/contacts-android#317] [Q&A] [answered]
I noticed that adding contacts with the library takes longer time compared to my previous implementation. Could it be that the library adds each individual contact with one query at a time instead of adding contacts in "chunks"?
According to my own research, you can queue up to 100 "ContentProviderOperation"'s at a time. So with my code I can add 10-15 contacts at a time, which speeds up the experience significantly.
As example my code for inserting multiple Contact's at once using "chunks":
```
val chunkedContacts = listOf(<My Chunked Contact's>, <My Chunked Contact's>, <My Chunked Contact's>)
chunkedContacts.forEach { chunkedC -> 
val inContacts = chunkedC
val contentProviderOperations = arrayListOf<ContentProviderOperation>()
        inContacts.forEach { contactDTO ->
            val rawId = contentProviderOperations.size
	    // 1. Create Contact with Account and Source_Id
            contentProviderOperations.add(
                ContentProviderOperation.newInsert(ContactsContract.RawContacts.CONTENT_URI)
                    .withValue(ContactsContract.RawContacts.ACCOUNT_TYPE, AccountUtils.ACCOUNT_TYPE)
                    .withValue(ContactsContract.RawContacts.ACCOUNT_NAME, AccountUtils.ACCOUNT_NAME)
                    .withValue(ContactsContract.RawContacts.SOURCE_ID, contactDTO.sourceId) // Add Unique identifier
                    //.withValue(ContactsContract.RawContacts.AGGREGATION_MODE, 				ContactsContract.RawContacts.AGGREGATION_MODE_DISABLED)
                    .build()
            )
	    // 2. Pepare Insert as IS_READ_ONLY
            val contentProviderOperationBuilder = ContentProviderOperation.newInsert(ContactsContract.Data.CONTENT_URI)
                .withValueBackReference(
                    ContactsContract.Data.RAW_CONTACT_ID,
                    rawId
                )
                .withValue(ContactsContract.Data.IS_READ_ONLY, 1) // Readonly == true
                .withValue(
                    ContactsContract.Data.MIMETYPE,
                    ContactsContract.CommonDataKinds.StructuredName.CONTENT_ITEM_TYPE
                )
	    // 3. Set Display Name if present
            if(!contactDTO.displayName.isNullOrBlank()){
                contentProviderOperationBuilder.withValue(
                    ContactsContract.CommonDataKinds.StructuredName.DISPLAY_NAME,
                    contactDTO.displayName
                )
            }else{
		// 3. Set Names
                contentProviderOperationBuilder//
                    .withValue(ContactsContract.CommonDataKinds.StructuredName.GIVEN_NAME, contactDTO.firstName)
                    .withValue(ContactsContract.CommonDataKinds.StructuredName.FAMILY_NAME, contactDTO.lastName)
                    .withValue(ContactsContract.CommonDataKinds.StructuredName.MIDDLE_NAME, contactDTO.middleName)
            }
            contentProviderOperations.add(contentProviderOperationBuilder.build())
            // 4. Add Organisation - Title
            if (!contactDTO.title.isNullOrBlank()) {
                contentProviderOperations.add(
                    ContentProviderOperation.newInsert(ContactsContract.Data.CONTENT_URI)
                        .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, rawId)
                        .withValue(ContactsContract.CommonDataKinds.Organization.MIMETYPE, ContactsContract.CommonDataKinds.Organization.CONTENT_ITEM_TYPE)
                        .withValue(ContactsContract.CommonDataKinds.Organization.TITLE, contactDTO.title)
                        .build());
            }
	    // 5. Add Organisation - Company
            if(!contactDTO.company.isNullOrBlank()){
                contentProviderOperations.add(
                    ContentProviderOperation.newInsert(ContactsContract.Data.CONTENT_URI)
                        .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, rawId)
                        .withValue(ContactsContract.CommonDataKinds.Organization.MIMETYPE, 				ContactsContract.CommonDataKinds.Organization.CONTENT_ITEM_TYPE)
                        .withValue(ContactsContract.CommonDataKinds.Organization.COMPANY, contactDTO.company)
                        .build())
            }
            // 6. Add Work number
            if (!contactDTO.otherNumber.isNullOrBlank()) {
                contentProviderOperations.add(
                    ContentProviderOperation.newInsert(ContactsContract.Data.CONTENT_URI)
                        .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, rawId)
                        .withValue(ContactsContract.Data.MIMETYPE, ContactsContract.CommonDataKinds.Phone.CONTENT_ITEM_TYPE)
                        .withValue(ContactsContract.CommonDataKinds.Phone.NUMBER, contactDTO.workNumber)
                        .withValue(ContactsContract.Data.MIMETYPE, ContactsContract.CommonDataKinds.Phone.CONTENT_ITEM_TYPE)
                        .withValue(ContactsContract.CommonDataKinds.Phone.TYPE, 				ContactsContract.CommonDataKinds.Phone.TYPE_WORK)
                        .build());
            }
            // 7. Add Mobile number
            if (!contactDTO.mobileNumber.isNullOrBlank()) {
                contentProviderOperations.add(
                    ContentProviderOperation.newInsert(ContactsContract.Data.CONTENT_URI)
                        .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, rawId)
                        .withValue(ContactsContract.Data.MIMETYPE, ContactsContract.CommonDataKinds.Phone.CONTENT_ITEM_TYPE)
                        .withValue(ContactsContract.CommonDataKinds.Phone.NUMBER, contactDTO.mobileNumber)
                        .withValue(ContactsContract.Data.MIMETYPE, ContactsContract.CommonDataKinds.Phone.CONTENT_ITEM_TYPE)
                        .withValue(ContactsContract.CommonDataKinds.Phone.TYPE, 				ContactsContract.CommonDataKinds.Phone.TYPE_MOBILE)
                        .build());
            }
            // 8. Add Other number
            if (!contactDTO.telephoneNumber.isNullOrBlank()) {
                contentProviderOperations.add(
                    ContentProviderOperation.newInsert(ContactsContract.Data.CONTENT_URI)
                        .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, rawId)
                        .withValue(ContactsContract.Data.MIMETYPE, ContactsContract.CommonDataKinds.Phone.CONTENT_ITEM_TYPE)
                        .withValue(ContactsContract.CommonDataKinds.Phone.NUMBER, contactDTO.otherNumber)
                        .withValue(ContactsContract.Data.MIMETYPE, ContactsContract.CommonDataKinds.Phone.CONTENT_ITEM_TYPE)
                        .withValue(ContactsContract.CommonDataKinds.Phone.TYPE, 				ContactsContract.CommonDataKinds.Phone.TYPE_OTHER)
                        .build());
            }
        }
	// contentProviderOperations should not have more than 100 Operations! otherwise Android will ignore them!
        try {
            val result = inContentResolver.applyBatch(
                ContactsContract.AUTHORITY,
                contentProviderOperations
            )
        } catch (inE: Exception) {
        }
}
```
I have numbered the "contentProviderOperations" (1. - 8.) to show that I have to do 8 operations to add a contact.
Since you can only add 100 contentProviderOperations at once, you would have to divide the list of contacts into chunks.
For me, this means that I can add 100 / 8 = 12.5 ~= 12 contacts at once. So I can chunk the list of contacts with the value 12.

**Comment by vestrel00** [comment]
Hey! Yes, currently the API inserts each RawContact and any associated data rows as part of a single batch of operations. You can see this in https://github.com/vestrel00/contacts-android/blob/main/core/src/main/java/contacts/core/Insert.kt

It looks something like this (pseudocode),

```Kotlin
for (rawContact in rawContacts) {
    results[rawContact] = contentResolver.applyBatch(insertRawContactAndData(rawContact))
}
```

This keeps the code clean and simple. One raw contact equals one batch of operations executed atomically by itself. 

Another benefit of doing it this way is that failure to insert one RawContact will not affect insertion of another RawContact. For example, if you are using the `Insert` API to insert two RawContacts, one RawContact insertion may fail and the other may succeed. You can see failures vs success in the `Result` returned.

I wrote the API this way because the most typical usage of insertions is when the end-user is creating a RawContact. Typically, this is done one at a time in a full-screen activity where the end-user can fill in RawContact data such as name, organization, etc...

I always try to balance code complexity vs performance WITH RESPECT TO the typical use case.

In this case, I agree that your own insertion code is more performant than this library because it can potentially insert multiple RawContacts in a single batch. However, I don't really see that being a common use case. Also, if someone really wanted to perform a mass insertion of RawContacts (maybe 1,000 RawContacts), then they can still use this library. They can chunk the insertions into separate threads/coroutines so that multiple insertions can occur at the same time. It is still not as performant as your own code but should do the trick. 

> So with my code I can add 10-15 contacts at a time, which speeds up the experience significantly.

Could you provide some real numbers, in milliseconds? How much faster are we talking about? Could you provide some benchmarks of inserting 15 RawContacts using your own code and also using this library?

When performing benchmarks, please insert the same amount of data per raw contact to keep the speed comparison fair. 

If the different in speed is less than 15ms (1ms per RawContact) or some arbitrary low number, then I do not see value in pursuing this discussion further. Always think about the ROI (return on investment) 😁

**Reply by marrale** [reply]
Thank you for the quick reply.
My app is designed for sync 2000+ contacts.
Tomorrow I will probably do the benchmarks tests (September 22nd) and I will let you know the test results.

**Reply by vestrel00** [reply]
Yes, I figured it was related to syncing large amounts of RawContacts =) 

Looking forward to the benchmark results! I'm sure the difference will be big because that is not a trivial number 😅 

I definitely think that the extra optimizations that I'll be making will help speed things up! The question will be, "will it be fast enough?"

How much of the extra internal checks/guards can I allow API users to disable in order to gain some performance whilst still keeping code complexity low... We will find out 😁

**Reply by marrale** [reply]
@vestrel00 I was able to do benchmark tests, these are the results by inserting 1000 contacts on empty device (it's a 5y old device):
With my own code > 25253 millis => 25 sec (Chunked 33)
By the library (no chunking) > 248628 millis => 248 sec
By the library (with 33 chunking) > 215034 millis => 215 sec
It gests a small improvement by chunking using the library, but the difference is huge between own code and library.

Here the used code for this benchmark, so you can do also some benchmark tests on your test device.
My own code:
```
val contacts = mutableListOf<ContactDTO>()
        for (i in 1 until inAmount + 1){
            contacts.add(
                ContactDTO(
                    firstName = "John $i",
                    lastName = "Doe $i",
                    middleName = "Middle $i"
                )
            )
        }

        val chunkedContacts = contacts.chunked(33)
        val measuredTimeInMillis = measureTimeMillis {
            chunkedContacts.forEach { chunk ->
                addMultipleContacts(contentResolver, chunk)
            }
        }
System.err.println("Own Insert Took: $measuredTimeInMillis millis == ${measuredTimeInMillis / 1000} sec")


suspend fun addMultipleContacts(inContentResolver: ContentResolver, inContactDTOs: List<ContactDTO>){
        val contentProviderOperations = arrayListOf<ContentProviderOperation>()
        inContactDTOs.forEach { contactDTO ->
            val rawId = contentProviderOperations.size
            contentProviderOperations.add(
                ContentProviderOperation.newInsert(ContactsContract.RawContacts.CONTENT_URI)
                    /*.withValue(ContactsContract.RawContacts.ACCOUNT_TYPE, ACCOUNT_TYPE)
                    .withValue(ContactsContract.RawContacts.ACCOUNT_NAME, ACCOUNT_NAME)*/
                    .withValue(ContactsContract.RawContacts.SOURCE_ID, contactDTO.id) // Add Unique identifier
                    //.withValue(ContactsContract.RawContacts.AGGREGATION_MODE, ContactsContract.RawContacts.AGGREGATION_MODE_DISABLED)
                    .build()
            )

            val contentProviderOperationBuilder = ContentProviderOperation.newInsert(
                ContactsContract.Data.CONTENT_URI)
                .withValueBackReference(
                    ContactsContract.Data.RAW_CONTACT_ID,
                    rawId
                )
                .withValue(ContactsContract.Data.IS_READ_ONLY, 1) // Readonly == true
                .withValue(
                    ContactsContract.Data.MIMETYPE,
                    ContactsContract.CommonDataKinds.StructuredName.CONTENT_ITEM_TYPE
                )


            contentProviderOperationBuilder//
                .withValue(ContactsContract.CommonDataKinds.StructuredName.GIVEN_NAME, contactDTO.firstName)
                .withValue(ContactsContract.CommonDataKinds.StructuredName.FAMILY_NAME, contactDTO.lastName)
                .withValue(ContactsContract.CommonDataKinds.StructuredName.MIDDLE_NAME, contactDTO.middleName)

            contentProviderOperations.add(contentProviderOperationBuilder.build())
        }

        try {
            val a = inContentResolver.applyBatch(
                ContactsContract.AUTHORITY,
                contentProviderOperations
            )
        } catch (inE: Exception) {
            System.err.println("Can't insert LocalPhone Contacts!: ${inE.message} >>> ${inE.cause} >>> $inE")
        }
    }
```

Insert by library code:
```
        val contactsApi = Contacts(applicationContext)
        val newRawContacts = mutableListOf<NewRawContact>()

        for (i in 1 until inAmount + 1){
            newRawContacts.add(NewRawContact(
                name = NewName(
                    givenName = "John $i",
                    familyName = "Doe $i",
                    middleName = "Middle $i"
                )
            ))
        }
        
        val measuredTime = measureTimeMillis {
                val insertResult = contactsApi
                    .insert()
                    .rawContacts(newRawContacts)
                    .commitWithContext()
                System.err.println("Insert Complete. Successful: ${insertResult.isSuccessful}")
        }
        System.err.println("Library Insert took: $measuredTime millis == ${measuredTime / 1000} sec")
```

Insert by library with chunking code:
```
val contactsApi = Contacts(applicationContext)
        val newRawContacts = mutableListOf<NewRawContact>()

        for (i in 1 until inAmount + 1){
            newRawContacts.add(NewRawContact(
                name = NewName(
                    givenName = "John $i",
                    familyName = "Doe $i",
                    middleName = "Middle $i"
                )
            ))
        }

        val chunkedRawContacts = newRawContacts.chunked(33)
        //val startTime = System.currentTimeMillis()
        val defferedList = mutableListOf<Deferred<Insert.Result>>()
        chunkedRawContacts.forEach { chunk ->
            defferedList.add(contactsApi
                .insert()
                .rawContacts(chunk)
                .commitAsync())
        }
        val measuredTime = measureTimeMillis {
            while(defferedList.any { d -> !d.isCompleted }){}
        }
        System.err.println("Library Insert took: $measuredTime millis == ${measuredTime / 1000} sec")
```

**Reply by vestrel00** [reply]
Library insertion code is 10x slower!!! There's definitely something going on here. I do hope the optimizations I'm making will be enough to get it down from ~250sec to something less than 100sec with the benchmark code you provided.

In the benchmark code, I see two things that are different;

1. Your own code is executed in the main thread? The library code is executed in a different coroutine/thread when using `commitWithContext` or `commitAsync`. The difference is probably trivial but for the benchmark, could you also just use `commit` so that there is no context/thread switching?
2. ~~Your own code uses `ContactsContract.RawContacts.AGGREGATION_MODE_DISABLED`, which the library does not. Depending on the results of benchmarking the upcoming alpha build, I may want to add the option to use that in the library too.~~ Oh, I see it's commented out in the benchmark code XD lol

Otherwise, the benchmark looks good and simple. I see that no group memberships are inserted, only name so that makes things even more simple 🙂

**Reply by vestrel00** [reply]
I will also run the same benchmark on my 3 yr old Samsung Galaxy A71 physical device and try out my optimizations before I give you the alpha build. 

If I find the results unsatisfactory, I'll also look into the refactor you are suggesting with your initial discussion comment. I can definitely do it, I just currently like how simple the code is right now and I think this will increase code complexity so I'm trying to avoid it, if possible. But, if I can increase insertion speed of a large number of RawContacts by 10x, I will definitely do it 🙂

Another thing I'm thinking of is to implement another commit function. Something like `commitInChunks()`. This way, I do not have to make refactors (no regression risk). The behavior is different in that it will be faster when inserting larger numbers of RawContacts BUT with the caveat that an error/exception caused by inserting one RawContact may cause other RawContact in the same chunk/batch to also fail.

Anyways, I'll keep you updated!

**Reply by vestrel00** [reply]
Okay, so... I'm running the benchmark on my laptop emulator because my 3yr old physical device takes too long to insert 1000 contacts lol. Here's what I've got.

## Insertion using this library

With all the optimizations I have made **SO FAR** to the `Insert` API of this library, running the following code,

```kotlin
private fun insertRawContacts() {
    contacts
        .insert()
        .rawContacts(createRawContacts())
        .allowBlanks(true)
        .validateAccounts(false)
        .validateGroupMemberships(false)
        .include()
        .commit()
}

private fun createRawContacts() = buildList {
    for (i in 1..1000) {
        add(
            NewRawContact(
                name = NewName(
                    givenName = "John $i",
                    middleName = "Middle $i",
                    familyName = "Doe $i"
                )
            )
        )
    }
}
```

Takes **on average 9-10 seconds**. Of course, this number should NOT be compared to the numbers you got since my laptop emulator is extremely fast. Much faster than high-end physical devices.

Note that I'm able to get the time lapse by looking at the logs, filtering with the tag `"ContactsDebug"`. You can do this by providing a `logger` when creating the `Contacts` API instance. For example,

```kotlin
val contactsApi = Contacts(context, logger = AndroidLogger(redactMessages = !BuildConfig.DEBUG))
```

## Insertion using custom code

As for insertion using `ContactsContract` directly (not this library), similar to your custom code,

```kotin
private fun insertRawContacts() {
    val startTime = System.currentTimeMillis()

    createRawContacts()
        .chunked(50) // 2 operation * 50 RawContacts chunk = 100 operations
        .forEach { rawContactsChunk ->
            val operations = arrayListOf<ContentProviderOperation>()
            rawContactsChunk.forEach {
                val rawContactIdOpIndex = operations.size
                operations.add(
                    ContentProviderOperation.newInsert(ContactsContract.RawContacts.CONTENT_URI)
                        .withValue(ContactsContract.RawContacts.ACCOUNT_NAME, null)
                        .withValue(ContactsContract.RawContacts.ACCOUNT_TYPE, null)
                        .build()
                )
                operations.add(
                    ContentProviderOperation.newInsert(ContactsContract.Data.CONTENT_URI)
                        .withValueBackReference(
                            ContactsContract.Data.RAW_CONTACT_ID,
                            rawContactIdOpIndex
                        )
                        .withValue(
                            ContactsContract.Data.MIMETYPE,
                            ContactsContract.CommonDataKinds.StructuredName.CONTENT_ITEM_TYPE
                        )
                        .withValue(
                            ContactsContract.CommonDataKinds.StructuredName.GIVEN_NAME,
                            it.name?.givenName
                        )
                        .withValue(
                            ContactsContract.CommonDataKinds.StructuredName.MIDDLE_NAME,
                            it.name?.middleName
                        )
                        .withValue(
                            ContactsContract.CommonDataKinds.StructuredName.FAMILY_NAME,
                            it.name?.familyName
                        )
                        .build()
                )
            }

            contentResolver.applyBatch(ContactsContract.AUTHORITY, operations)
        }

    val endTime = System.currentTimeMillis()
    Log.d("YOLO", "Took ${endTime - startTime} milliseconds")
}

private fun createRawContacts() = buildList {
    for (i in 1..1000) {
        add(
            NewRawContact(
                name = NewName(
                    givenName = "John $i",
                    middleName = "Middle $i",
                    familyName = "Doe $i"
                )
            )
        )
    }
}
```

In the above code, I set a max number of operations per batch to 100 (like what you are doing).

Takes **on average 2 seconds**. 

## Conclusion

There is clearly a significant performance gain when chunking operations of multiple RawContacts!

Performance increases as the number of data per RawContact decreases. It makes sense because we are able to insert larger numbers of RawContacts per batch when there are only 2 operations per RawContact compared to 8 operations per RawContact (100 divided by a smaller number equates to more RawContacts inserted per batch). Performance gain does not exist if each RawContact contains 99 data items since then that would equate to 1 RawContact inserted per batch.

## Next steps

I will create an issue for adding a new commit function for the `Insert` API. I will probably name it `commitInChunks()` (similar naming to `commitInOneTransaction`). 

I will then perform the benchmark using the new commit function and post the results in this thread 😁

**Reply by vestrel00** [reply]
I created the issue! #329 

I will be including it in the next release ([0.3.2](https://github.com/vestrel00/contacts-android/discussions/316)) 🔥 

Again, thanks so much for raising this issue! This library will be that much better with the next release 😄

**Reply by vestrel00** [reply, marked as answer]
@marrale, I'm excited to tell you that I have pushed up the alpha tag; [`0.3.2-alpha1`](https://github.com/vestrel00/contacts-android/releases/tag/0.3.2-alpha1) !! 

Please feel free to use it. Just note that I will not publish a formal release for it. Everything in that tag and perhaps more will be incorporated in the upcoming [0.3.2 release](https://github.com/vestrel00/contacts-android/discussions/316).

Please use the following code snippet to insert large quantities of RawContacts in the fastest way possible when using this library,

```kotlin
contactsApi
    .insert()
    .rawContacts(newRawContacts)
    .allowBlanks(true)
    .validateAccounts(false)
    .validateGroupMemberships(false)
    .include()
    .commitInChunks()
```

Apart from the other optimizations I made, I also implemented your suggestion. Please use `commitInChunks` instead of `commit` when inserting large amounts of RawContacts. There is a caveat with `commitInChunks` that you need to be aware of, which is documented at https://vestrel00.github.io/contacts-android/basics/insert-contacts/#commitinchunks-vs-commit

-----------------

I performed benchmarks once again on my physical Samsung Galaxy A71 (API 33). Here are the results 😀 

|                                                        | Execution time |
|--------------------------------------------------------|----------------|
| custom code                                            | 12.2 seconds   |
| 0.3.1 (`commit`)                                       | 72.5 seconds   |
| 0.3.2-alpha1 (`commit` w/ other optimizations)         | 37.8 seconds   |
| 0.3.2-alpha1 (`commitInChunks` w/ other optimizations) | 12.5 seconds   |

Note that I ran each of these benchmarks around 10 times each. One thing I did notice is that _0.3.2-alpha1 (`commitInChunks` w/ other optimizations_ is pretty much the same as the _custom code_! Although, on average, the _custom code_ seem to beat out the library by a few hundred milliseconds with 1,000 RawContacts sample size. However, there were some runs where the _custom code_ actually performed a few hundred milliseconds slower than the library! 

Anyways, I think the key takeaway here is that you should now be able to insert large quantities of RawContacts at the same, or extremely similar, speeds as your custom code! But, I'll let you be the judge of that 😉 

@marrale, please let me know how it performs on your end. I'm hoping that it good enough for your use case 🤞

**Please mark this as the answer if it satisfies your request** 😁

**Reply by marrale** [reply]
@vestrel00 sorry for late reply, I'm not at home right now. Thanks for implementing the chunked insertion in the library and have build the alpha version. Monday I will be back and for sure I have time to test the alpha version also on my device with my test app. I will then add the benchmark results to this thread.

**Reply by marrale** [reply]
@vestrel00 I have done the benchmark tests on our test device with the new alpha library and the performance now is very good, similar to our custom code. The difference with validation or whiteout is small. Thanks a lot to have done this improvements.

**Reply by vestrel00** [reply]
Great to hear!!! Awesome 🤩

**Comment by vestrel00** [comment]
I also changed the title of the discussion to focus on just "insert" because you didn't mention deletion anywhere. 

Furthermore, the `Delete` API in this library has a function called `commitInOneTransaction` that will allow you to perform deletion of several contacts in a single batch; https://github.com/vestrel00/contacts-android/blob/0.3.1/core/src/main/java/contacts/core/Delete.kt#L210

This `commitInOneTransaction` function exist for some delete APIs in the library but not in update and insert APIs because it was cheap to implement for delete APIs (code complexity is low). Also, deletion of several contacts is much much more common end-user use-case (e.g. selecting multiple contacts to delete in a list/recyclerview) than insertion of several contacts.

**Comment by vestrel00** [comment]
Aside from batching several different RawContacts in a single batch operation, I do see room for speed improvements when inserting RawContacts.

Currently, for each RawContact, the `Insert` API will query all accounts in the system to make sure that the provided account exist. If not, then it will coerce it to null; https://github.com/vestrel00/contacts-android/blob/0.3.1/core/src/main/java/contacts/core/Insert.kt#L424

Also, if there are GroupMemberships data to be inserted, the API will also internally perform several queries on Accounts and Groups to make sure that the memberships are valid (belonging to groups that actually exist for the particular chosen Account and that there are no duplicate memberships); https://github.com/vestrel00/contacts-android/blob/0.3.1/core/src/main/java/contacts/core/Insert.kt#L477

I could provide library users an option to disable those guards (but default to enabled) in order to speed up insertion of multiple RawContacts (still in separate batch operations).

The two things I mentioned above might be more of a contributing factor to why you are seeing a noticeable difference in speed in your code vs this library (more than because I'm performing inserts of RawContacts and associated data in separate batches). I'll create issues for these later today and work on them ASAP. 

Once I complete the issues, I will create an alpha tag of the library so that you can try out the new functions to see if that helps improve insertion speeds for your use case 😁

**Reply by vestrel00** [reply]
Apart from those extra internal queries I mentioned, the API also checks if a particular column/property should be inserted with a value depending on the list of included fields. So, there is a `includedFields.contains(field)` for every RawContact data. That could also be a factor in speed, though probably not as much as the Accounts and Groups queries I mentioned. 

There are several other small things like that that are written to support certain API features and also guard against bad/unnecessary input data such as blank values. All of these things can add up to a performance hit.

Here is an exhaustive list of extra things that the `Insert` API does internally upon calling `commit()`;

1. checks permissions
2. checks if operation is cancelled
3. check if RawContact is blank (no data)
4. checks if Account is valid (in the system)
5. checks if group memberships are valid (point to groups that belong to the specified Account)
6. checks if certain fields are included
7. checks if data is blank

All of the above cost some CPU. Some more than others. Currently, you can already disable (3) by setting `allowBlanks(true)`. I will be creating issues to give users the ability to disable (4), (5), and (6). As for (1), (2), and (7), IMO those should not be allowed to be disabled (performance hit should be trivial anyways). Hopefully, by disabling (3, 4, 5, and 6), you will see a "good-enough" increase in speed 🤞

When building an API such as this, there is a balancing act that must occur between quantity, quality, and code complexity of features VERSUS speed/performance. I am hoping to give you more speed/performance without sacrificing other things.

Anyways, look forward to the issues I'll be creating and an alpha tag for you to try out. I appreciate you calling out these potential improvements to this library!

**Reply by vestrel00** [reply]
I created the following issues, which will be included in the next release ([0.3.2](https://github.com/vestrel00/contacts-android/discussions/316));

- #318 
- #320 
- #321 

Once I have completed them (hopefully by end of this week), I will create an alpha tag (probably named `0.3.2-alpha1`) so that @marrale can try out the new optimizations. I hope that the extra optimizations will decrease insertion time to an acceptable level. 

Of course, your own custom code will always be more performant because it is written specifically for your use case. The APIs provided in this library have extra features that may be useful to some but not others. So, the best I can do is provide ways to opt-in or opt-out of those features.

If the extra optimizations is still not enough to get insertion speeds to an acceptable level, then I can revisit your original idea of performing a single batched operation for multiple RawContacts (instead of a single batched operation for every RawContact).

**Reply by marrale** [reply]
Once you have the alpha version I'm happy to do the speed tests with disabled (4), (5), and (6).
Thanks again for your commitment to keep optimizing the API.

**Reply by vestrel00** [reply]
Hey, #321 is taking a lot longer than I expected. I won't have the alpha version ready until maybe Monday. Sorry!🙏

**Reply by marrale** [reply]
Hi, no problem, I still successfully use the library for all other operations apart for the insert one ;-)

**Reply by vestrel00** [reply]
I have completed [all the issues I mentioned](https://github.com/vestrel00/contacts-android/discussions/317#discussioncomment-7064110). However, I may not be able to get an alpha release out on Monday because I'm digging deeper into this. Let's streamline this conversation by continuing it in this thread; https://github.com/vestrel00/contacts-android/discussions/317#discussioncomment-7089693
